### ISSUE TITLE
Span.h static asserts on MSVC 2017 with C++14/17 enabled.

### ISSUE THREAD

**Issue by acdemiralp** [VcDevel/Vc#239]
I have built and linked the library (version 1.4) through cmake, setting the premade flags. Note that the issue disappears when C++11 is explicitly set. The errors are:
```cxx
Schweregrad	Code	Beschreibung	Projekt	Datei	Zeile	Unterdrückungszustand
Fehler	C2338	Can't have a span with an extent < 0	prs	c:\program files\vc\include\vc\common\span.h	143	
Fehler	C2118	Negativer Index	prs	c:\program files\vc\include\vc\common\span.h	167	
Fehler	C2572	"Vc_1::Common::span<const std::byte,-1>::span": Neudefinition des Standardarguments: Parameter 1	prs	c:\program files\vc\include\vc\common\span.h	204	
Fehler	C2382	"Vc_1::Common::span<const std::byte,-1>::span": Neudefinition; unterschiedliche Ausnahmespezifikationen	prs	c:\program files\vc\include\vc\common\span.h	204	
Fehler	C2572	"Vc_1::Common::span<std::byte,-1>::span": Neudefinition des Standardarguments: Parameter 1	prs	c:\program files\vc\include\vc\common\span.h	204	
Fehler	C2382	"Vc_1::Common::span<std::byte,-1>::span": Neudefinition; unterschiedliche Ausnahmespezifikationen	prs	c:\program files\vc\include\vc\common\span.h	204	
```

**Comment by acdemiralp** [comment]
Problem lies in the conditionals for the `__cpp_lib_byte`. Taking them out allows compilation:
```
#ifdef __cpp_lib_byte
    span<const std::byte, dynamic_extent> _as_bytes() const noexcept
    {
        return {reinterpret_cast<const std::byte*>(data()), size_bytes()};
    }

    span<std::byte, dynamic_extent> _as_writeable_bytes() const noexcept
    {
        return {reinterpret_cast<std::byte*>(data()), size_bytes()};
    }
#endif  // __cpp_lib_byte
```

**Comment by markstock** [comment]
I am seeing this problem as well. Visual Studio 15.9 2017 on Windows 10, hardware is old Phenom II 965 X4, Vc using master from 2019-06-25.

Vc builds just fine for "x64" platform. The above error appears when building my own software which uses Vc (and C++17). As @acdemiralp states, removing those lines in `span.h` does allow the build to proceed.

Now I am stuck during linking, but that's unrelated to this.

**Comment by tgeijten** [comment]
I'm seeing the same issue with Visual Studio 16.5.4. However, I still get the same error after removing the `#ifdef` / `#endif` conditionals. Any suggestions..?

**Comment by anders-wind** [comment]
Have anyone found a solution for this? Hopefully one that does not include editing the headers of Vc? 
So far I can see it only fails to compile if i set c++ standard to 17 but not to 14.

**Comment by bernhardmgruber** [comment]
I nailed the issue down to a bug in MSVC and reported it here: https://developercommunity2.visualstudio.com/t/wrong-instantiation-of-primary-template-while-pars/1296700 (see link to godbolt).

The fix will be released with MSVC 16.10. That could solve the issue.

**Comment by EricAtORS** [comment]
@bernhardmgruber 
I was having the same issue.
I checked out the compiler explorer code you linked and it now compiles if you set the compiler to `msvc v19.latest`.

Thanks for reporting the issue to microsoft!

**Comment by bernhardmgruber** [comment]
You are welcome! Let's see if the entire issue in Vc is resolved with MSVC 16.10 then. The bug I reported could only be part of the compilation failure.
